### PR TITLE
Add in GoogleTest support

### PR DESCRIPTION
**Pull request by zebmason**

### CODE REVIEW

**Comment by zebmason** [comment]
Cleaned up my code and placed in a new branch which avoid all the CMake on CI issues which I tried to resolve using
1. [GitHub actions](https://github.com/marketplace/actions/get-cmake) but couldn't figure out the dependencies.
2. Downloading the binary of CMake and setting the path failed on Ubuntu Bionic for some unknown reason.
3. Downloading the source of CMake and building led to dependency hell

**Comment by zebmason** [comment]
@weshinsley Thanks for merging in #331 earlier, this has been reworked to add in GoogleTest support purely for CMake builds

**Comment by zebmason** [comment]
@weshinsley rebased

**Comment by weshinsley** [comment]
Is it essential to move the VS project files? Just that all the team here use those for developing and will have them open while they pull... (not a big problem if it is essential...!)

**Comment by zebmason** [comment]
@weshinsley  Sorry it seems this has suffered from bit rot over time, I thought I'd removed all the vcxproj changes but they seem to have made it back in. Perhaps due to rebasing.

File: CMakeLists.txt [hunk 1]
```diff
@@ -52,8 +52,10 @@ endif()
 # Specify the C++ standard
 set(CMAKE_CXX_STANDARD 14)
 
+add_subdirectory(third_party)
 add_subdirectory(src)
 
+add_subdirectory(unit_tests)
```
Comment (matt-gretton-dann) [inline CMakeLists.txt:58]: Move this lower down so you don't need the `enable_testing()` statement in unit_tests/CMakeLists.txt

File: CMakeLists.txt [hunk 2]
```diff
@@ -52,8 +52,10 @@ endif()
...
 if(WIN32 AND USE_WIN32_BMP)
   message(WARNING "USE_WIN32_BMP is enabled.  Therefore testing is disabled.")
   message(WARNING "To enable testing invoke with `cmake -DUSE_WIN32_BMP=NO`")
```
Comment (matt-gretton-dann) [inline CMakeLists.txt:61]: Memo to self: This is no longer needed and should be removed in a separate PR.

File: unit_tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+include_directories("${gtest_SOURCE_DIR}/include")
+
+enable_testing()
```
Comment (matt-gretton-dann) [inline unit_tests/CMakeLists.txt:3]: Remove (see CMakeLists.txt comment).

File: unit_tests/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+
+ADD_SUBDIRECTORY(noddy)
```
Comment (matt-gretton-dann) [inline unit_tests/CMakeLists.txt:5]: Boringly: Can we the noddy directory and Holder.cpp so that we don't have a Noddy Holder reference please?

File: unit_tests/noddy/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+FIND_PACKAGE(Threads REQUIRED)
+
+# from list of files we'll create tests test_name.cpp -> test_name
+foreach(_test_file ${TEST_NODDY_SRC_FILES})
+    get_filename_component(_test_name ${_test_file} NAME_WE)
+    add_executable(${_test_name} ${_test_file})
+    target_link_libraries(${_test_name} gtest_main ${CMAKE_THREAD_LIBS_INIT})
+    add_test(${_test_name} ${_test_name})
+endforeach()
```
Comment (matt-gretton-dann) [inline unit_tests/noddy/CMakeLists.txt:13]: Instead of `add_test` can we use `gtest_discover_tests`?  This improves the potentials for parallelism?

File: unit_tests/CMakeLists.txt (hunk 4 above)
Comment (zebmason) [reply unit_tests/CMakeLists.txt:5]: Sorry I don't understand, there seems to be a word missing

Comment (weshinsley) [reply unit_tests/CMakeLists.txt:5]: I think the missing word is probably *rename* ... We might have to keep the Noddy Holder joke private to ourselves...!

File: CMakeLists.txt (hunk 1 above)
Comment (zebmason) [reply CMakeLists.txt]: Done

File: unit_tests/CMakeLists.txt (hunk 3 above)
Comment (zebmason) [reply unit_tests/CMakeLists.txt]: Done

File: unit_tests/CMakeLists.txt (hunk 4 above)
Comment (zebmason) [reply unit_tests/CMakeLists.txt:18]: Renamed to Noddy Test

File: unit_tests/noddy/CMakeLists.txt (hunk 5 above)
Comment (zebmason) [reply unit_tests/noddy/CMakeLists.txt]: Done, but removed the third party directory so that all the Google Test stuff is pulled into the correct directory. So have done a force push rather than add and remove.

File: unit_tests/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+if(NOT googletest_POPULATED)
+  FetchContent_Populate(googletest)
+  set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
+  add_subdirectory(${googletest_SOURCE_DIR} ${googletest_BINARY_DIR})
+endif()
+include(GoogleTest)
+
+include_directories("${gtest_SOURCE_DIR}/include")
+
+ADD_SUBDIRECTORY(noddy)
```
Comment (matt-gretton-dann) [inline unit_tests/CMakeLists.txt:18]: Sorry we need to get change Noddy as well...  (It refers to a character created by Enid Blyton).

Comment (zebmason) [reply unit_tests/CMakeLists.txt]: Done

**Comment by weshinsley** [comment]
(Ref Issue #233)
